scripts/codemods/typed_flags.py

```diff
@@ -0,0 +1,182 @@
+        if self.context.full_module_name != "disnake.flags":
+            raise codemod.SkipFile("this module contains no definitions of flag classes.")
+
+        # we preformulate a list of all flag classes on the imported flags module
+        all_flag_classes = []
+        for attr_name in dir(flags):
+            obj = getattr(flags, attr_name)
+            if (
+                isinstance(obj, type)
+                and issubclass(obj, BASE_FLAG_CLASSES)
+                and obj not in BASE_FLAG_CLASSES
+            ):
+                all_flag_classes.append(obj.__name__)
+        self.flag_classes = all_flag_classes
+
+        return super().transform_module(tree)
+
+    def visit_ClassDef(self, node: cst.ClassDef) -> Optional[bool]:
+        # no reason to continue into classes
+        return False
+
+    def leave_ClassDef(self, _: cst.ClassDef, node: cst.ClassDef):
+        if not m.matches(node.name, m.OneOf(*map(m.Name, self.flag_classes))):
+            return node
+
+        # now the meat of the function
+        # we have two options here, and both of them depend on if __init__ exists, and *how* __init__ exists.
+        # thankfully, we aren't limited to just codemodding, we can import the module if we need it
+        # if __init__ exists without a TYPE_CHECKING block, we create and add two overloads
+        # if __init__ exists under a TYPE_CHECKING block, we remove and replace it
+        # we also check and remove previously existing TYPE_CHECKING and overloads
+
+        # and we have to import the module anyways, to get a list of valid flags for each subclass.
+        flag: flags.BaseFlags = getattr(flags, node.name.value)
+        all_flags = sorted(flag.VALID_FLAGS.keys())
+
+        # determine which method to use for creating the init overload, and possibly the update overload too.
+        # only a few of these need to be overloads, in which we can use the same technique from permission overloads.
+        # for the other ones, we need to find a TYPE_CHECKING block that *may* or may not contain a function definition.
+        # if none of the TYPE_CHECKING blocks contain a function definition, we make a new one and
+        # insert it near the beginning of the class body.
+        # we also decorate with @_generated so we can delete it later.
+
+        hide_behind_typechecking: bool = True
+        if_block: Optional[cst.If] = None
+        init: Optional[cst.FunctionDef] = None
+        body = list(node.body.body)
+        kwonly_params = [
+            cst.Param(cst.Name(flag_name), cst.Annotation(cst.Name("bool")), default=cst.Ellipsis())
+            for flag_name in all_flags
+        ]
+        for b in body.copy():
+            if m.matches(b, m.FunctionDef(m.Name("__init__"))) and isinstance(b, cst.FunctionDef):
+                deco_names = [getattr(deco.decorator, "value", "") for deco in b.decorators]
+                if "_generated" in deco_names:
```

Comment:
```suggestion
                if any(getattr(deco.decorator, "value", "") == "_generated" for deco in b.decorators):
```
could perhaps even make this a matcher?

Reply:
https://github.com/DisnakeDev/disnake/pull/778/commits/efea6aec9e4dd0d62e4d1a9278339dc53026f01db